Remotion/Web/ClientScript/Scripts/ExecutionEngine.ts

```diff
@@ -0,0 +1,280 @@
+    {
+      if (args.Status === this._httpStatusFunctionTimeout)
+      {
+        if (window.console)
+          window.console.warn ('WXE function has timed out. Stopping refresh- and abort-requests.');
+
+        if (this._refreshTimer !== null)
+        {
+          window.clearInterval (this._refreshTimer);
+          this._refreshTimer = null;
+        }
+        this._isAbortEnabled = false;
+      }
+    };
+    SmartPage_Context.Instance!.SendOutOfBandRequest(this._refreshUrl + '&WxePage_Garbage=' + Math.random(), successHandler, errorHandler);
+  };
+
+  // Evaluates whether the postback request should continue.
+  // returns: true to continue with request
+  public CheckFormState (isAborting: boolean, hasSubmitted: boolean, hasUnloaded: boolean, isCached: boolean): boolean
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isAborting', isAborting);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('hasSubmitted', hasSubmitted);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('hasUnloaded', hasUnloaded);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isCached', isCached);
+
+    if (this._isCacheDetectionEnabled
+        && (isCached || hasSubmitted || hasUnloaded))
+    {
+      this.ShowStatusIsCachedMessage();
+      return false;
+    }
+    if (isAborting)
+    {
+      this.ShowStatusIsAbortingMessage();
+      return false;
+    }
+    else
+    {
+      return true;
+    }
+  };
+
+  // Shows the "page is aborting" message
+  public ShowStatusIsAbortingMessage(): void
+  {
+    if (this._statusIsAbortingMessage != null)
+      SmartPage_Context.Instance!.ShowMessage('WxeStatusIsAbortingMessage', this._statusIsAbortingMessage);
+  };
+
+  // Shows the "page is cached" message
+  public ShowStatusIsCachedMessage(): void
+  {
+    if (this._statusIsCachedMessage != null)
+      SmartPage_Context.Instance!.ShowMessage('WxeStatusIsCachedMessage', this._statusIsCachedMessage);
+  };
+
+  public EstablishPageKeepAliveLock = function(): void
+  {
+    var lockName = btoa("" + (Math.random() * 10e18));
+    var lockIndex = 0;
+
+    function onLockTaken(lock: Lock)
+    {
+      if (lock === null)
+      {
+        lockIndex++;
+        navigator.locks!.request(lockName + lockIndex, { ifAvailable: true }, onLockTaken);
+        return null;
+      }
+      else
+      {
+        return new Promise(function (resolve, reject) { });
+      }
+    }
+
+    // The navigator.locks API is only available for secure origins, i.e. sites using HTTPS.
+    // The navigator.locks API is only available in Chromium-based browsers, versions 69.* and later.
+    if (navigator.locks && navigator.locks.request)
+    {
+      navigator.locks.request(lockName + lockIndex, { ifAvailable: true }, onLockTaken);
+    }
+  }
+
+  // The single instance of the WxePage_Context object
+  private static _instance: Nullable<WxePage_Context> = null;
+
+  public static GetInstance(): Nullable<WxePage_Context>
+  {
+    return WxePage_Context._instance;
```

Comment:
Depends on if you want to explicitly handle the case where `_instance === null`. e.g. throw an exception here. If we assume that this should not be null because you call init before accessing anything else, we can fix it inline

Reply:
Nah, I prefer this to be null-safe. I don't wan to have any more bangs than necessary so all our own APIs should be null-typed correctly.